sensibo/unified.js

```diff
@@ -1,235 +1,485 @@
+	 */
+	getInternalAcState: function (device) {
+		const internalSmartModeState = this.getInternalSmartModeState(device)
+		const internalFilterState = this.getInternalFilterState(device)
+		const internalSwingState = this.getInternalSwingState(device)
+		const fanSpeed = this.getFanSpeed(device)
+		const resultingInternalAcState = new Classes.InternalAcState(
+			device.acState.on,
+			device.acState.mode.toUpperCase(),
+			device.acState.targetTemperature ? device.acState.temperatureUnit === 'C' ? device.acState.targetTemperature : this.toCelsius(device.acState.targetTemperature) : null,
+			device.measurements.temperature,
+			device.measurements.humidity,
+			internalSmartModeState,
+			device.acState.light && device.acState.light !== 'off',
+			device.pureBoostConfig && device.pureBoostConfig.enabled,
+			internalFilterState.filterChange,
+			internalFilterState.filterLifeLevel,
+			internalSwingState.horizontalSwing,
+			internalSwingState.verticalSwing,
+			fanSpeed
+		)
+
+		return resultingInternalAcState
+	},
 
-		state.airQuality = device.measurements?.pm25 ?? 0
+	/**
+	 * @param {import('../types').Device} device
+	 * @param {SensiboACPlatform} platform
+	 * @returns {Classes.InternalAirPurifierState}
+	 */
+	getInternalAirPurifierState: function (device, platform) {
+		const internalSmartModeState = this.getInternalSmartModeState(device)
+		const internalFilterState = this.getInternalFilterState(device)
+		const internalSwingState = this.getInternalSwingState(device)
+		const fanSpeed = this.getFanSpeed(device)
+		const airQualityState = this.getAirQualityState(device, platform)
+		const resultingInternalAirPurifierState = new Classes.InternalAirPurifierState(
+			device.acState.on,
+			device.acState.mode.toUpperCase(),
+			device.acState.targetTemperature ? device.acState.temperatureUnit === 'C' ? device.acState.targetTemperature : this.toCelsius(device.acState.targetTemperature) : null,
+			device.measurements.temperature,
+			device.measurements.humidity,
+			internalSmartModeState,
+			device.acState.light && device.acState.light !== 'off',
+			device.pureBoostConfig && device.pureBoostConfig.enabled,
+			internalFilterState.filterChange,
+			internalFilterState.filterLifeLevel,
+			internalSwingState.horizontalSwing,
+			internalSwingState.verticalSwing,
+			fanSpeed,
+			airQualityState.airQuality,
+			airQualityState.VOCDensity,
+			airQualityState.carbonDioxideDetected,
+			airQualityState.carbonDioxideLevel
+		)
+
+		return resultingInternalAirPurifierState
+	},
 
-		if (device.measurements?.tvoc && device.measurements.tvoc > 0) {
-			// convert ppb to μg/m3
-			const VOCDensity = Math.round(device.measurements.tvoc * 4.57)
+	/**
+	 * @param {import('../types').Device} device
+	 * @param {SensiboACPlatform} platform
+	 * @returns {Classes.InternalAirQualitySensorState}
+	 */
+	getAirQualityState: function (device, platform) {
+		// convert ppb to μg/m3
+		let VOCDensity = Math.round(device.measurements.tvoc * 4.57)
```

Comment:
```suggestion
		let VOCDensity = Math.round(device.measurements?.tvoc * 4.57)
```

Or you could add a function wide check that device.measurements is set and remove the various null coalesce checks